src/NHibernate/Cache/ReadOnlyCache.cs

```diff
@@ -11,34 +11,39 @@ namespace NHibernate.Cache
 	/// </summary>
 	public partial class ReadOnlyCache : IBatchableCacheConcurrencyStrategy
 	{
+#pragma warning disable 618
 		private ICache cache;
-		private IBatchableReadOnlyCache _batchableReadOnlyCache;
-		private IBatchableCache _batchableCache;
+#pragma warning restore 618
+		private CacheBase _batchableCache;
 		private static readonly INHibernateLogger log = NHibernateLogger.For(typeof(ReadOnlyCache));
 
 		/// <summary>
 		/// Gets the cache region name.
 		/// </summary>
 		public string RegionName
 		{
-			get { return cache.RegionName; }
+			get { return Cache.RegionName; }
 		}
 
+		// 6.0 TODO: type as CacheBase instead and switch to autoproperty
+#pragma warning disable 618
 		public ICache Cache
+#pragma warning restore 618
 		{
 			get { return cache; }
 			set
 			{
 				cache = value;
-				// ReSharper disable once SuspiciousTypeConversion.Global
-				_batchableReadOnlyCache = value as IBatchableReadOnlyCache;
-				_batchableCache = value as IBatchableCache;
+				_batchableCache = value as CacheBase ?? new ObsoleteCacheWrapper(value);
```

Comment:
Is there any value in having 2 fields to hold a cache and a wrapper?

Reply:
It avoids to get through the wrapper for cases not requiring it, but it is likely to get inlined by compilation optimizations. So not really.